mm2src/coins/qrc20/history.rs

```diff
@@ -0,0 +1,991 @@
+pub struct UtxoFromHashFuture {
+    future: Option<Box<dyn Future<Output = Result<UtxoTx, String>> + Unpin + 'static>>,
+}
+
+pub enum HistoryOrder {
+    NewestToOldest,
+    OldestToNewest,
+}
+
+impl HistoryBuilder {
+    pub fn new(coin: Qrc20Coin) -> HistoryBuilder {
+        let address = qrc20_addr_from_utxo_addr(coin.utxo.my_address.clone());
+        let token_address = coin.contract_address;
+        HistoryBuilder {
+            coin,
+            from_block: 0,
+            topic: QRC20_TRANSFER_TOPIC.to_string(),
+            address,
+            token_address,
+            order: None,
+        }
+    }
+
+    #[allow(clippy::wrong_self_convention)]
+    pub fn from_block(mut self, from_block: i64) -> HistoryBuilder {
+        self.from_block = from_block;
+        self
+    }
+
+    #[allow(dead_code)]
+    pub fn topic(mut self, topic: &str) -> HistoryBuilder {
+        self.topic = topic.to_string();
+        self
+    }
+
+    pub fn address(mut self, address: H160) -> HistoryBuilder {
+        self.address = address;
+        self
+    }
+
+    #[allow(dead_code)]
+    pub fn token_address(mut self, token_address: H160) -> HistoryBuilder {
+        self.token_address = token_address;
+        self
+    }
+
+    pub fn order(mut self, order: HistoryOrder) -> HistoryBuilder {
+        self.order = Some(order);
+        self
+    }
+
+    pub async fn build(self) -> Result<HistoryCont<TxHistoryItem>, String> {
+        self.build_with_rpc_error().await.map_err(|e| ERRL!("{}", e))
+    }
+
+    pub async fn build_with_rpc_error(self) -> Result<HistoryCont<TxHistoryItem>, JsonRpcError> {
+        let electrum = match self.coin.utxo.rpc_client {
+            UtxoRpcClientEnum::Electrum(ref rpc_cln) => rpc_cln,
+            UtxoRpcClientEnum::Native(_) => panic!("Native mode doesn't support"),
+        };
+
+        let address = qrc20_addr_into_rpc_format(&self.address);
+        let token_address = qrc20_addr_into_rpc_format(&self.token_address);
+        let mut history = electrum
+            .blockchain_contract_event_get_history(&address, &token_address, &self.topic)
+            .compat()
+            .await?;
+
+        if self.from_block != 0 {
+            history = history
+                .into_iter()
+                .filter(|item| self.from_block <= item.height)
+                .collect();
+        }
+
+        match self.order {
+            Some(HistoryOrder::NewestToOldest) => {
+                history.sort_unstable_by(|a, b| sort_newest_to_oldest(a.height, b.height))
+            },
+            Some(HistoryOrder::OldestToNewest) => {
+                history.sort_unstable_by(|a, b| sort_oldest_to_newest(a.height, b.height))
+            },
+            None => (),
+        }
+
+        Ok(HistoryCont { history })
+    }
+
+    /// Request history by `tx_hash` and wrap it into list of UtxoLazyFuture.
+    /// This method is used when there is no reason to load not all `UtxoTx`,
```

Comment:
Looks like it's meant that `This method is used when there is no reason to load all UtxoTx`, because otherwise there's double negative usage in sentence.